source/extensions/common/tap/admin.cc

```diff
@@ -110,40 +109,201 @@ void AdminHandler::unregisterConfig(ExtensionConfig& config) {
+    // Check if we already responded to the client
+    if (trace_buffer->flushed()) {
+      return;
+    }
+    // Main thread dispatcher serializes access to the trace_buffer
+    trace_buffer->bufferTrace(shared_trace);
+    // If the trace buffer is not full yet, wait to buffer more traces
+    if (!trace_buffer->full()) {
       return;
     }
 
-    std::string output_string;
-    switch (format) {
-      PANIC_ON_PROTO_ENUM_SENTINEL_VALUES;
-    case envoy::config::tap::v3::OutputSink::JSON_BODY_AS_STRING:
-    case envoy::config::tap::v3::OutputSink::JSON_BODY_AS_BYTES:
-      output_string = MessageUtil::getJsonStringFromMessageOrError(*trace, true, true);
-      break;
-    case envoy::config::tap::v3::OutputSink::PROTO_BINARY:
-      PANIC("not implemented");
-    case envoy::config::tap::v3::OutputSink::PROTO_BINARY_LENGTH_DELIMITED:
-      PANIC("not implemented");
-    case envoy::config::tap::v3::OutputSink::PROTO_TEXT:
-      PANIC("not implemented");
+    std::vector<envoy::data::tap::v3::TraceWrapper> buffer = trace_buffer->flush();
+
+    ENVOY_LOG(debug, "admin writing buffered trace list to response");
+
+    // Serialize writes to the stream
+    for (size_t i = 0; i < buffer.size(); i++) {
+      // Close stream on final message
+      attached_request->streamMsg(buffer[i], (i + 1) == buffer.size());
     }
+  });
+}
+
+AdminHandler::AttachedRequest::AttachedRequest(AdminHandler* admin_handler,
+                                               const envoy::admin::v3::TapRequest& tap_request,
+                                               Server::AdminStream* admin_stream)
+    : config_id_(tap_request.config_id()), config_(tap_request.tap_config()),
+      admin_stream_(admin_stream), main_thread_dispatcher_(admin_handler->main_thread_dispatcher_) {
+}
 
-    ENVOY_LOG(debug, "admin writing buffered trace to response");
-    Buffer::OwnedImpl output_buffer{output_string};
-    parent.attached_request_.value().admin_stream_->getDecoderFilterCallbacks().encodeData(
-        output_buffer, false);
+AdminHandler::AttachedRequestBuffered::AttachedRequestBuffered(
+    AdminHandler* admin_handler, const envoy::admin::v3::TapRequest& tap_request,
+    Server::AdminStream* admin_stream)
+    : AttachedRequest(admin_handler, tap_request, admin_stream) {
+  auto sink = tap_request.tap_config().output_config().sinks()[0];
+
+  uint64_t max_buffered_traces = sink.buffered_admin().max_traces();
+  uint64_t timeout = PROTOBUF_GET_MS_OR_DEFAULT(sink.buffered_admin(), timeout, 0);
+  trace_buffer_ = std::make_unique<TraceBuffer>(max_buffered_traces);
+  // Start the countdown if provided an actual timeout
+  if (timeout > 0) {
+    timer_ = dispatcher().createTimer(
+        [this, admin_handler] { this->onTimeout(admin_handler->attached_request_); });
+    timer_->enableTimer(std::chrono::milliseconds(timeout));
+  }
+}
+
+void AdminHandler::AttachedRequestBuffered::onTimeout(
+    std::weak_ptr<AttachedRequest> weak_attached_request) {
```

Comment:
I haven't worked that much with weak_ptr but should this be passed by const ref?